### PR TITLE
ARROW-11744: [C++] Add xsimd dependency

### PR DESCRIPTION
**Pull request by pitrou**
xsimd is header-only and is bundled from a git changeset.

### CODE REVIEW

**Comment by pitrou** [comment]
There is a test failure on ARM: https://travis-ci.com/github/pitrou/arrow/jobs/485572758#L3045
@guyuqi @cyb70289 Do you want to help on this?

**Comment by pitrou** [comment]
Looking at the source, it seems the implementation of `xsimd::batch_bool<int8_t, 16>::any` may be buggy: https://github.com/xtensor-stack/xsimd/blob/master/include/xsimd/types/xsimd_neon_bool.hpp#L685-L689

**Comment by pitrou** [comment]
(in any case, looks like boolean reduction with Neon is tricky: https://stackoverflow.com/questions/31197216/optimizing-horizontal-boolean-reduction-in-arm-neon)

**Comment by pitrou** [comment]
Looks like a possible fix for xsimd could be:
```diff
diff --git a/include/xsimd/types/xsimd_neon_bool.hpp b/include/xsimd/types/xsimd_neon_bool.hpp
index 0d2baf2..75bf3e1 100644
--- a/include/xsimd/types/xsimd_neon_bool.hpp
+++ b/include/xsimd/types/xsimd_neon_bool.hpp
@@ -684,8 +684,15 @@ namespace xsimd
 
             static bool any(const batch_type& rhs)
             {
+#if XSIMD_ARM_INSTR_SET >= XSIMD_ARM8_64_NEON_VERSION
+                return vmaxvq_u8(rhs) != 0;
+#else
                 uint8x8_t tmp = vorr_u8(vget_low_u8(rhs), vget_high_u8(rhs));
-                return vget_lane_u8(vpmax_u8(tmp, tmp), 0);
+                tmp = vpmax_u8(tmp, tmp);
+                tmp = vpmax_u8(tmp, tmp);
+                tmp = vpmax_u8(tmp, tmp);
+                return vget_lane_u8(tmp, 0);
+#endif
             }
         };
     }
```

(a similar fix is required for `all()`)

**Comment by pitrou** [comment]
@edponce You might want to take a short look at this.

**Comment by pitrou** [comment]
Ok, I submitted a fix for the `any()` issue in xsimd: https://github.com/xtensor-stack/xsimd/pull/419

**Comment by cyb70289** [comment]
Add xsimd to /LICENSE.txt?

File: cpp/cmake_modules/ThirdpartyToolchain.cmake [hunk 1]
```diff
@@ -1904,6 +1919,30 @@ if(ARROW_WITH_RAPIDJSON)
   include_directories(SYSTEM ${RAPIDJSON_INCLUDE_DIR})
 endif()
 
+macro(build_xsimd)
+  message(STATUS "Building xsimd from source")
+  set(XSIMD_PREFIX "${CMAKE_CURRENT_BINARY_DIR}/rapidjson_ep/src/rapidjson_ep-install")
```
Comment (kou) [inline cpp/cmake_modules/ThirdpartyToolchain.cmake:1924]:
```suggestion
  set(XSIMD_PREFIX "${CMAKE_CURRENT_BINARY_DIR}/xsimd_ep/src/xsimd_ep-install")
```

**Comment by cyb70289** [comment]
> ```diff
> +#if XSIMD_ARM_INSTR_SET >= XSIMD_ARM8_64_NEON_VERSION
> +                return vmaxvq_u8(rhs) != 0;
> +#else
> ```

For Arm64, the `#if` should be true. Otherwise, current code will already fail Travis Arm64 CI.
So looks some macros are not correctly set in xsimd, or maybe should be set manually.

File: cpp/src/arrow/util/utf8.h [hunk 2]
```diff
@@ -240,54 +244,26 @@ inline bool ValidateAsciiSw(const uint8_t* data, int64_t len) {
...
       data += 32;
       data2 += 32;
       len -= 32;
     }
 
-    or1 = _mm_or_si128(or1, or2);
-    if (_mm_movemask_epi8(_mm_cmplt_epi8(or1, _mm_set1_epi8(0)))) {
+    // To test for upper bit in all bytes, test whether any of them is negative
+    or1 |= or2;
+    if (xsimd::any(or1 < zero)) {
```
Comment (cyb70289) [inline cpp/src/arrow/util/utf8.h:266]: Compared with current Neon code, this line introduces an additional vector comparison instruction. No trouble as it's not in the hot loop.
Look simd lib does make the code easier to read and maintain.

Comment (pitrou) [reply cpp/src/arrow/util/utf8.h:266]: If we contribute horizontal max to xsimd, we can probably remove the additional instruction as well.

**Comment by pitrou** [comment]
> Add xsimd to /LICENSE.txt?

I don't think so. AFAIK, LICENSE.txt is for code directly in the repository, not code that we may download at compile-time. @kou Am I wrong?

**Comment by pitrou** [comment]
Proof-of-concept bit-packing generation rewrite here, with Neon path (no sweat!): https://github.com/pitrou/arrow/compare/ARROW-11744-xsimd...pitrou:xsimd-bpacking

File: cpp/cmake_modules/ThirdpartyToolchain.cmake [hunk 3]
```diff
@@ -1634,14 +1645,18 @@ macro(build_gtest)
     GTEST_MAIN_SHARED_LIB
     "${_GTEST_LIBRARY_DIR}/${CMAKE_SHARED_LIBRARY_PREFIX}gtest_main${_GTEST_LIBRARY_SUFFIX}"
     )
+  set(GTEST_INSTALL_NAME_DIR "$<INSTALL_PREFIX$<ANGLE-R>/lib")
+  # Fix syntax highlighting mess introduced by unclosed bracket above
```
Comment (bkietz) [inline cpp/cmake_modules/ThirdpartyToolchain.cmake:1649]: :cry:

File: cpp/cmake_modules/ThirdpartyToolchain.cmake [hunk 4]
```diff
@@ -1904,6 +1919,30 @@ if(ARROW_WITH_RAPIDJSON)
...
+  add_dependencies(toolchain xsimd_ep)
+  add_dependencies(toolchain-tests xsimd_ep)
+
+  set(XSIMD_VENDORED TRUE)
+endmacro()
+
+# For now xsimd is always bundled from upstream
+build_xsimd()
+# TODO: Don't use global includes but rather target_include_directories
+include_directories(SYSTEM ${XSIMD_INCLUDE_DIR})
```
Comment (bkietz) [inline cpp/cmake_modules/ThirdpartyToolchain.cmake:1944]: For symmetry with the other libraries we could write:
```suggestion
# For now xsimd is always bundled from upstream
if(1)
  set(XSIMD_DEPENDENCY_SOURCE "BUNDLED")
  resolve_dependency(XSIMD)
  # TODO: Don't use global includes but rather target_include_directories
  include_directories(SYSTEM ${XSIMD_INCLUDE_DIR}) 
endif()
```

**Comment by kou** [comment]
> > Add xsimd to /LICENSE.txt?
> 
> I don't think so. AFAIK, LICENSE.txt is for code directly in the repository, not code that we may download at compile-time.

Right.

https://infra.apache.org/licensing-howto.html#step-by-step

> Do not modify LICENSE or NOTICE for non-bundled dependencies.

**Comment by cyb70289** [comment]
Benchmark xsimd against current intrinsic code. Build with clang-9.

For x86, there's 10% decrease of `ValidateLargeAscii`, may deserve some investigation.
For arm, the decrease of `ValidateSmallAscii` is expected due to an additional instruction.

```
skylake, master
-----------------------------------------------------------------------------------
Benchmark                         Time             CPU   Iterations UserCounters...
-----------------------------------------------------------------------------------
ValidateTinyAscii              3.87 ns         3.87 ns    180903250 bytes_per_second=2.40796G/s
ValidateSmallAscii             6.54 ns         6.54 ns    107085259 bytes_per_second=19.51G/s
ValidateLargeAscii             1655 ns         1655 ns       422102 bytes_per_second=56.266G/s

skylake, xsimd
-----------------------------------------------------------------------------------
Benchmark                         Time             CPU   Iterations UserCounters...
-----------------------------------------------------------------------------------
ValidateTinyAscii              3.87 ns         3.87 ns    181010171 bytes_per_second=2.40758G/s
ValidateSmallAscii             5.92 ns         5.92 ns    118181224 bytes_per_second=21.5656G/s
ValidateLargeAscii             1799 ns         1799 ns       388569 bytes_per_second=51.7653G/s
```

```
arm64, master
-----------------------------------------------------------------------------------
Benchmark                         Time             CPU   Iterations UserCounters...
-----------------------------------------------------------------------------------
ValidateTinyAscii              6.02 ns         6.02 ns    101048537 bytes_per_second=1.54793G/s
ValidateSmallAscii             9.68 ns         9.67 ns     72363717 bytes_per_second=13.1908G/s
ValidateLargeAscii             2162 ns         2161 ns       323625 bytes_per_second=43.0956G/s

arm64, xsimd
-----------------------------------------------------------------------------------
Benchmark                         Time             CPU   Iterations UserCounters...
-----------------------------------------------------------------------------------
ValidateTinyAscii              6.35 ns         6.35 ns     93579235 bytes_per_second=1.46676G/s
ValidateSmallAscii             12.7 ns         12.7 ns     55227661 bytes_per_second=10.068G/s
ValidateLargeAscii             2164 ns         2162 ns       323364 bytes_per_second=43.0724G/s
```

**Comment by cyb70289** [comment]
OK, re-tested with gcc-7.5. Performance is almost the same for xsimd and intrinsic code.
I think xsimd does not introduce performance penalty. We needn't care too much about the micro-benchmark.


```
skylake, master
-----------------------------------------------------------------------------------
Benchmark                         Time             CPU   Iterations UserCounters...
-----------------------------------------------------------------------------------
ValidateTinyAscii              4.89 ns         4.89 ns    143229610 bytes_per_second=1.90621G/s
ValidateSmallAscii             6.92 ns         6.92 ns    101370712 bytes_per_second=18.4492G/s
ValidateLargeAscii             1702 ns         1702 ns       411766 bytes_per_second=54.7354G/s

skylake, xsimd
-----------------------------------------------------------------------------------
Benchmark                         Time             CPU   Iterations UserCounters...
-----------------------------------------------------------------------------------
ValidateTinyAscii              4.65 ns         4.65 ns    148053313 bytes_per_second=2.00206G/s
ValidateSmallAscii             6.22 ns         6.22 ns    112602376 bytes_per_second=20.5258G/s
ValidateLargeAscii             1699 ns         1699 ns       412266 bytes_per_second=54.8192G/s
```

```
arm64, master
-----------------------------------------------------------------------------------
Benchmark                         Time             CPU   Iterations UserCounters...
-----------------------------------------------------------------------------------
ValidateTinyAscii              5.71 ns         5.71 ns     99296752 bytes_per_second=1.62991G/s
ValidateSmallAscii             7.61 ns         7.60 ns     92056037 bytes_per_second=16.7808G/s
ValidateLargeAscii             2130 ns         2130 ns       328973 bytes_per_second=43.7367G/s

arm64, xsimd
-----------------------------------------------------------------------------------
Benchmark                         Time             CPU   Iterations UserCounters...
-----------------------------------------------------------------------------------
ValidateTinyAscii              5.70 ns         5.69 ns    104548080 bytes_per_second=1.63584G/s
ValidateSmallAscii             7.69 ns         7.69 ns     91047673 bytes_per_second=16.6009G/s
ValidateLargeAscii             2127 ns         2127 ns       329007 bytes_per_second=43.7892G/s
```

**Comment by cyb70289** [comment]
+1 from me

**Comment by pitrou** [comment]
Ok, I'll merge this since there seems to be no objection. Thanks for the reviews.

**Comment by pitrou** [comment]
CI is green on my fork: https://github.com/pitrou/arrow/commit/44d4153eba81949b178550c240bf2dfe09efa3e2

**Comment by pitrou** [comment]
cc @JohanMabille :-)

**Pull request closed without merging**